### ISSUE TITLE
Inconsistency in loss on SAME data for train and validation modes

### ISSUE THREAD

**Issue by charlielito** [tensorflow/tensorflow#44568]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux (Colab)
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.3.0
- Python version: 3.6.9
- CUDA/cuDNN version: Colab
- GPU model and memory: Colab has k80 I think

**Describe the current behavior**
I'm implementing a semantic segmentation model with images. As a good practice I tested my training pipeline with just one image and tried to over-fit that image. To my surprise, when training with the exactly the same images, the loss goes to 0 as expected but when evaluating THE SAME IMAGES, the loss is much much higher, and it keeps going up as the training continues. So the segmentation output is garbage when `training=False`, but when run with `training=True` is works perfectly.

To be able to anyone to reproduce this I took the official [segmentation tutorial](https://www.tensorflow.org/tutorials/images/segmentation
) and modified it a little for training a convnet from scratch and just 1 image. The model is very simple, just a sequence of Conv2D with batch normalization and Relu. The results are the following

![Screenshot from 2020-11-03 18-04-17](https://user-images.githubusercontent.com/8033598/98050621-18cf3c80-1e00-11eb-93b0-0470054ddc50.png)

As you see the loss and eval_loss are really different, and making inference to the image gives perfect result in training mode and in eval mode is garbage.

**Describe the expected behavior**
I know Batchnormalization behaves differently in inference time since it uses the averaged statistics calculated whilst training. Nonetheless, since we are training with just 1 same image and evaluating in the same image, this shouldn't happen right? Moreover I implemented the same architecture with the same optimizer in Pytorch and this does not happen there. With pytorch it trains and eval_loss converges to train loss

**Standalone code to reproduce the issue**
Here you can find the above mentioned https://colab.research.google.com/drive/18LipgAmKVDA86n3ljFW8X0JThVEeFf0a#scrollTo=TWDATghoRczu
and at the end also the Pytorch implementation

**Comment by amahendrakar** [comment]
@charlielito,
I do not have access to the Colab Notebook you have linked. Could you please provide the required permissions to view the files. Thanks!

**Comment by charlielito** [comment]
@amahendrakar  sorry, forgot to change the permissions. Now you should be able to access it

**Comment by charlielito** [comment]
@amahendrakar

**Comment by amahendrakar** [comment]
@charlielito,
Sorry for the delayed response. 

@ymodak,
I was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/0fad546e9be6cb2836ed033cdeee99cf/44568.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/18d074388becf3d26924f8a1a0156be3/44568-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by aliFrancis** [comment]
I have found a very similar issue. Could this be linked in any way to the various discussions about Batch Normalization from 2018? e.g. http://blog.datumbox.com/the-batch-normalization-layer-of-keras-is-broken/ and https://github.com/keras-team/keras/pull/9965

My understanding was that this had been fixed in TF2, but looks like a very similar problem, relating to differences in behaviour of batch norm with `training' value set to True or False.

**Comment by francescomilano172** [comment]
I'm also having a similar issue. However, I noticed that in the colab file above, if we set the number of epochs `EPOCHS` to a larger value (e.g., 400), the validation loss will eventually also decrease steadily and converge to 0 like the train loss. I suspect that, while there is definitely something to be checked in the way the training parameters of the `BatchNormalization` layer are learned (and possibly initialized as well), the problem might not necessarily be caused by a "logical" bug in the `tf.keras` implementation. Indeed, by the way the batch-normalization layer works:
- In training mode, the layer will try to learn parameters that best fit the statistics of the training data (a single image in the case above), but these learned parameters are only used in *inference* mode. In training mode, the forward pass will use the statistics of the actual input data rather than the learned parameters. This means that **it's possible that at the same time the network overfits perfectly to the training data, but the learned parameters from the batch normalization layer do not represent the statistics of the input perfectly**. This is coherent with the fact when the network has overfitted, running in training mode its prediction is perfect: the learned parameters of the batch normalization layers will be ignored (the statistics from the input data will be used instead), while the other layers will use the parameters optimized for the data to which it has overfitted.
- In inference mode, the layer will actually use the learned parameters rather than the input statistics. This is also coherent with the wrong prediction from the overfitted network: my claim is that at epoch 100 the learned parameters of batch normalization do not represent well the statistics of the training data yet, but they gradually do so later (epoch 400).

I believe that the problem might be related also to the number of other learnable parameters, and that the network could somehow learn to overfit these parameter "faster" than the batch-normalization parameters. As a quick test, I reduced the number of `ConvBNReLU` layers in the model from 15 to 2, and the validation loss is still larger than the training loss at the beginning, but way closer to it than when 15 layers are used. Furthermore, it does not increase over time, and also eventually converges to 0.

To conclude, I believe that this is a major issue, since in practice I would expect the statistics of large datasets to take (much) longer to be learned than those from this toy example, but it could be that the layer is still doing "what it should do".

**Comment by charlielito** [comment]
It might be an idiomatic issue, but I think what you denote as _learned parameters_ are not actually _learned_ parameters (learned in sense of being updated by backprop), but just the exponential averaged statistics from the training batches (mean and var). 

Batch norm __does__ have learned parameters which are _beta_ and _alpha_, which are used indeed in botch training and testing modes. Just the normalization (mean and var) in training are calculated from the batch data whilst in inference from the values averaged while training. See https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization

So I thought, as you said, it had something to do with the averaged statistics not converging fast enough to the real statistics . Because we are training with 1 image and testing with the same image I denote as _real_ statistics the statistics from that image. If you set `momentum=0.0` in the BatchNorm layer, the averaged statistics should match perfectly with the statistics from the current batch (which is just 1 image, `moving_mean = moving_mean * momentum + mean(batch) * (1 - momentum)`). But even so you see that the validation loss is still very high. So the problem is something  or somewhere else

**Comment by francescomilano172** [comment]
Great! Yes, by _learnable_ parameters I meant the "non-trainable" parameters `self.moving_mean` and `self.moving_var` which are updated with the data received during training. And good to see that it was indeed too high a `momentum`that was causing slow convergence, at least now we know where the problem was.
